### ISSUE TITLE
nested strings are not allowed outside of a11y string object for key: keyboardHelpDialog.grabOrReleaseBall

### ISSUE THREAD

**Issue by samreid** [phetsims/center-and-variability#522]
During a build, we got this error:

```
~/phet/root/center-and-variability$ grunt --brands=phet
Running "lint-all" task
[........................................] 100.00%

Running "report-media" task

Running "clean" task

Running "build" task
>> tsc complete: 3709ms
Building runnable repository (center-and-variability, brands: phet)
Building brand: phet
>> Webpack build complete: 2985ms
Fatal error: Perennial task failed:
AssertionError [ERR_ASSERTION]: nested strings are not allowed outside of a11y string object for key: keyboardHelpDialog.grabOrReleaseBall
    at Object.getStringEntryFromMap (/Users/samreid/phet/root/chipper/js/common/ChipperStringUtils.js:156:63)
    at /Users/samreid/phet/root/chipper/js/grunt/getStringMap.js:218:46
    at Array.forEach (<anonymous>)
    at /Users/samreid/phet/root/chipper/js/grunt/getStringMap.js:213:15
    at Array.forEach (<anonymous>)
    at /Users/samreid/phet/root/chipper/js/grunt/getStringMap.js:212:23

```

We were recently instructed by @jessegreenberg that the keyboard help strings should be translatable, so we cannot nest them under the `a11y` key. However, do we need to flatten everything so it can show in rosetta?

**Comment by jessegreenberg** [comment]
Good catch, yes I think they need to be flattened to support translation.

**Comment by samreid** [comment]
OK thanks!

**Comment by samreid** [comment]
More comments from slack:

```
Sam Reid
  [17 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693424872010129)
[@jessegreenberg](https://phetsims.slack.com/team/U6ELY17N3)
 and 
[@jbphet](https://phetsims.slack.com/team/U6E5GM3RB)
 or others who are aware. We were instructed to get rid of the “a11y” key in the keyboard help strings for https://github.com/phetsims/center-and-variability/issues/316 but now the sim fails during build, saying nested strings are not allowed outside of a11y string object for key: keyboardHelpDialog.grabOrReleaseBall which is in https://github.com/phetsims/center-and-variability/issues/522.  So should we flatten all of them? There are quite a lot.
21 replies


Jesse Greenberg
  [17 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693424926490329?thread_ts=1693424872.010129&cid=C6HPE0J91)
Sorry I forgot about that! Yes, my understanding is that they must be flattened if they are to be translatable at the moment.


Sam Reid
  [16 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693424993691119?thread_ts=1693424872.010129&cid=C6HPE0J91)
OK. Just camel case it all? keyboardHelpDialog.medianScreen.moveGrabbedBallOrCard => keyboardHelpDialogMedianScreenMoveGrabbedBallOrCard?  Will we regret that soon?


Michael Kauzmann
  [14 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425071047009?thread_ts=1693424872.010129&cid=C6HPE0J91)
can you get away with just making it a string with a dot in it?
'keyboardHelpDialog.medianScreen.moveGrabbedBallOrCard'
:+1::skin-tone-2:
1



Sam Reid
  [14 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425101259789?thread_ts=1693424872.010129&cid=C6HPE0J91)
Wow, that’s great if it works.


Michael Kauzmann
  [14 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425114647179?thread_ts=1693424872.010129&cid=C6HPE0J91)
that is how we support non a11y "nesting'" at this time.


Michael Kauzmann
  [13 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425127487809?thread_ts=1693424872.010129&cid=C6HPE0J91)
and you still access it in the code with object dot notation


John Blanco
:office:  [12 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425190665009?thread_ts=1693424872.010129&cid=C6HPE0J91)
Yes, you can put dots in string keys.  I fear that this may haunt us one day, but for now it is possible and is done elsewhere. (edited) 


Sam Reid
  [11 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425248265219?thread_ts=1693424872.010129&cid=C6HPE0J91)
[@jbphet](https://phetsims.slack.com/team/U6E5GM3RB)
 is that what you recommend for this release?


Michael Kauzmann
  [11 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425268630129?thread_ts=1693424872.010129&cid=C6HPE0J91)
we punted on rosetta supporting the nesting, so this is probably the best solution for now. We would likely write an easy script to "nest" dot-notationed strings in the english string files, so not too problematic here.
:+1:
1



Michael Kauzmann
  [11 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425299543959?thread_ts=1693424872.010129&cid=C6HPE0J91)
Lots of precedent for this btw https://github.com/phetsims/joist/blob/7544630a4f25946d7d2356d4f1043ddfd4d6bccc/joist-strings_en.json#L7-L24


John Blanco
:office:  [10 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425332637249?thread_ts=1693424872.010129&cid=C6HPE0J91)
Yes, that's what you should do.


John Blanco
:office:  [9 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425395437729?thread_ts=1693424872.010129&cid=C6HPE0J91)
Will this create a bunch of new strings that will need to be translated in existing translations?


John Blanco
:office:  [9 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425410932789?thread_ts=1693424872.010129&cid=C6HPE0J91)
Or is it specific to C&V?


Michael Kauzmann
  [9 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425415438969?thread_ts=1693424872.010129&cid=C6HPE0J91)
new sim


Michael Kauzmann
  [9 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425418937679?thread_ts=1693424872.010129&cid=C6HPE0J91)
not common code


John Blanco
:office:  [8 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425426135729?thread_ts=1693424872.010129&cid=C6HPE0J91)
Good, cool, thanks.


Michael Kauzmann
  [8 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425443911239?thread_ts=1693424872.010129&cid=C6HPE0J91)
We only put things under a11y if we DONT want them translatable 
[@samreid](https://phetsims.slack.com/team/U6DKKJ2FN)
. BTW
:+1:
1



Sam Reid
  [7 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425490335459?thread_ts=1693424872.010129&cid=C6HPE0J91)
C&V, not common code, but C&V is published as a prototype with a lot of strings. But no keyboard help dialog in the prototype--it’s new for this upcoming version. (edited) 





Chris Malley
  [7 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425497806989?thread_ts=1693424872.010129&cid=C6HPE0J91)
Whatever you do here, make sure you keep Studio discoverability of StringProperty in mind.
:+1:
1



Chris Malley
  [7 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425520249739?thread_ts=1693424872.010129&cid=C6HPE0J91)
… which is why I’m not wild about a flat 'keyboardHelpDialog.medianScreen.moveGrabbedBallOrCard' (edited) 


Sam Reid
  [2 minutes ago](https://phetsims.slack.com/archives/C6HPE0J91/p1693425804928269?thread_ts=1693424872.010129&cid=C6HPE0J91)
Wow GPT-4 just restructured that in about 10 seconds. Thanks for making my life easier GPT-4!
```

**Comment by samreid** [comment]
Fixed, thanks everyone! Closing.

**Comment by samreid** [comment]
After this change (using flat keys with dots), they are still nested correctly in studio.
